xcm/pallet-xcm-benchmarks/src/generic/benchmarking.rs

```diff
@@ -0,0 +1,355 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use crate::{new_executor, worst_case_holding, XcmCallOf};
+use codec::Encode;
+use frame_benchmarking::{benchmarks, BenchmarkError};
+use frame_support::dispatch::GetDispatchInfo;
+use sp_std::vec;
+use xcm::{
+	latest::{prelude::*, MultiAssets},
+	DoubleEncoded,
+};
+
+const MAX_ASSETS: u32 = 100;
+
+benchmarks! {
+	query_holding {
+		let holding = worst_case_holding();
+
+		let mut executor = new_executor::<T>(Default::default());
+		executor.holding = holding;
+
+		let instruction = Instruction::<XcmCallOf<T>>::QueryHolding {
+			query_id: Default::default(),
+			dest: T::valid_destination()?,
+			assets: Wild(WildMultiAsset::All), // TODO is worst case filter?
+			max_response_weight: u64::MAX,
+		};
+
+		let xcm = Xcm(vec![instruction]);
+
+	} : {
+		executor.execute(xcm)?;
+	} verify {
+		// The assert above is enough to validate this is completed.
+		// todo maybe XCM sender peek
+	}
+
+	// This benchmark does not use any additional orders or instructions. This should be managed
+	// by the `deep` and `shallow` implementation.
+	buy_execution {
+		let holding = worst_case_holding();
+
+		let mut executor = new_executor::<T>(Default::default());
+		executor.holding = holding;
+
+		let fee_asset = Concrete(Here.into());
```

Comment:
This may be the same case as `QueryHolding` in that the weight depends on what kind of assets that you put here. It ultimately calls [`Assets::general_take`](https://github.com/paritytech/polkadot/blob/f26005b999acb8936bda66df10daa0d75c85ce62/xcm/xcm-executor/src/assets.rs#L244-L324) to match the wildcard against the holding register to see which ones to take out for weight payment.

Reply:
For all the chains we care about right now, the only way you can pay fees is with the base currency

Reply:
Okay, we may need to revisit sometime in the future when NFTs can be used to pay for weights, e.g. coupons.